### DISCUSSION TITLE
Hacking newer TSUN TSOL M-800 (DE) with wifi (really MS800 with s/n R17E..)

### DISCUSSION THREAD

**Discussion by benderl** [tbnobody/OpenDTU#855] [Ideas]
I have no Idea, if this is the right place to discuss, as OpenDTU is specialized in using NRF transmitters. So please just close this if it of no interest here.

Yesterday I got hands on a TSOL M-800(DE) with wifi embedded. Unfortunately the s/n is completely different to the ones described in the tutorials: R17E...
This new inverter has wifi embedded and you have to use an app from TSUN to enter yout SSID and password. So far, so bad. As I had problems creating an account yesterday, the inverter is still disconnected. Time to fiddle around.

I noticed a new bluetooth device with the s/n as device name. So lets try to connect and see, what can be done. Currently I am limited to a Pi for testing as the inverter is in a different building some kilometers away. Anyway, the pi reported the same device I noticed yesterday on my phone. On the cli I was able to connect running `bluetoothctl` in interactive mode. I am not quite shure, if the connection is stable, as the pi is about 10m apart from the inverter and there is a wall in the way. Sometimes I get disconnected after a while but this may be related to poor quality of the link.

This is what `bluetoothctl info` tells me about the device:
```
# info
Device 90:EB:48:xx:xx:xx (public)
	Name: R17Exxxxxxxxxxxx
	Alias: R17Exxxxxxxxxxxx
	Appearance: 0x0341
	Paired: no
	Trusted: yes
	Blocked: no
	Connected: yes
	LegacyPairing: no
	UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
	UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
	UUID: Amazon.com Services, Inc. (0000fe00-0000-1000-8000-00805f9b34fb)
```

I tried service discovery with `sdptool`, but the tool could not detect sdp service. As I am a newbie with bluetooth on cli, I would be very glad to get some help here.

Tomorrow I will be able to connect the inverter with the wifi network. Maybe then I can further investigate on the network side.

**Comment by tbnobody** [comment]
The mentioned inverter should be one of these: https://priwatt.de/wordpress/wp-content/uploads/2023/03/User-manual_GEN3-series-Microinverter_MS800MS400_V1.1_230213.pdf

The specification shows: `built-in WiFi module; 4G module (Optional)`
That would mean, it's definetly  not compatible with OpenDTU as it does not use either the NRF Module or the 868 MHz CMT2300A module


Just read the pdf some pages further... on page 22 you see some example serial numbers... They all begin with R17E.... That means you are using a MSxxx inverter

**Reply by stefan123t** [reply]
The DTUBI with built-in WiFi DTU opens a Network AccessPoint called `DTUBI-<Serialnumber>` and advertises itself via NetworkProtocol using NetworkCommands on IP `10.10.100.254` Port `10081`. We are in the process of analysing the protocol and suggesting a feasible approach to implement this along the already supported NRF24L01+ and CMT2300A based Gen2 (MI) and Gen3 (HM, HMS/HMT) Protocols.
See https://github.com/lumapu/ahoy/issues/1189 for reference.

**Comment by benderl** [comment]
Ok, thanks for the link. Yes, it's definitely one of this series of inverters. It was advertised as M800(DE), but now I have the real product name for searching.

I think this discussion can be closed.

**Comment by LenzGr** [comment]
@benderl I stumbled over this thread during my research on how to communicate with this device directly - did you have any success? I aim to integrate the collected metrics into Home Assistant.

**Comment by benderl** [comment]
Sorry, had no luck with this inverter.

**Comment by stefan123t** [comment]
You can both check #1430 were the repo with a proof-of-concept for reading the protobuf based Gen3 S-Cloud API / SDK App API by Henk Wiedig is linked.

**Comment by sulimanfarzat** [comment]
please see the folwing bolgs:
https://easywatt24.com/blogs/balkonkraftwerke/tsun
https://easywatt24.com/blogs/balkonkraftwerke/tsun-wechselrichter-entdrosseln-sei-bereit-fur-800-watt-per-over-the-air-update

**Reply by stefan123t** [reply]
@sulimanfarzat thanks for the links. 

Did anyone download the APKs and took a stab at decompiling them with JADX or something similar ?
